hack/lib/util/docs.sh

```diff
@@ -0,0 +1,139 @@
+			generated=$(grep -v "^${skipprefix}" "${tmpdir}/${file}") || :
+			if [[ "${original}" == "${generated}" ]]; then
+				# overwrite generated with original.
+				mv "${dest}/${file}" "${tmpdir}/${file}"
+			fi
+		else
+			rm "${dest}/${file}" || true
+		fi
+	done <"${dest}/${filestore}"
+
+	# put the new generated file into the destination
+	find "${tmpdir}" -exec rsync -pt {} "${dest}" \; >/dev/null
+	#cleanup
+	rm -rf "${tmpdir}"
+
+	echo "Assets generated in ${dest}"
+}
+readonly -f generate_manual_pages
+
+function generate_documentation() {
+	local dest="$1"
+	local skipprefix="${1:-}"
+
+	# We do this in a tmpdir in case the dest has other non-autogenned files
+	# We don't want to include them in the list of gen'd files
+	local tmpdir="${OS_ROOT}/_tmp/gen_doc"
+	mkdir -p "${tmpdir}"
+	# generate the new files
+	gendocs "${tmpdir}"
+	# create the list of generated files
+	ls "${tmpdir}" | LC_ALL=C sort > "${tmpdir}/.files_generated"
+
+	# remove all old generated file from the destination
+	while read file; do
+		if [[ -e "${tmpdir}/${file}" && -n "${skipprefix}" ]]; then
+			local original generated
+			original=$(grep -v "^${skipprefix}" "${dest}/${file}") || :
+			generated=$(grep -v "^${skipprefix}" "${tmpdir}/${file}") || :
+			if [[ "${original}" == "${generated}" ]]; then
+				# overwrite generated with original.
+				mv "${dest}/${file}" "${tmpdir}/${file}"
+			fi
+		else
+			rm "${dest}/${file}" || true
+		fi
+	done <"${dest}/.files_generated"
+
+	# put the new generated file into the destination
+	find "${tmpdir}" -exec rsync -pt {} "${dest}" \; >/dev/null
+	#cleanup
+	rm -rf "${tmpdir}"
+
+	echo "Assets generated in ${dest}"
+}
+readonly -f generate_documentation
+
+# os::util::gen-docs generates docs and manpages for the all the binaries
+# created for Origin.
+function os::util::gen-docs() {
+	os::util::ensure::built_binary_exists 'gendocs'
+	os::util::ensure::built_binary_exists 'genman'
+
+	OUTPUT_DIR_REL=${1:-""}
+	OUTPUT_DIR="${OS_ROOT}/${OUTPUT_DIR_REL}/docs/generated"
+	MAN_OUTPUT_DIR="${OS_ROOT}/${OUTPUT_DIR_REL}/docs/man/man1"
+
+	mkdir -p "${OUTPUT_DIR}" || echo $? > /dev/null
+	mkdir -p "${MAN_OUTPUT_DIR}" || echo $? > /dev/null
+
+	generate_documentation "${OUTPUT_DIR}"
+	generate_manual_pages "${MAN_OUTPUT_DIR}" "oc"
+	generate_manual_pages "${MAN_OUTPUT_DIR}" "openshift"
+	generate_manual_pages "${MAN_OUTPUT_DIR}" "oadm"
+}
+readonly -f os::util::gen-docs
+
+# os::util::set-man-placeholder puts a placeholder for every generated manpage.
+function os::util::set-man-placeholder() {
+	MAN_OUTPUT_DIR="$1"
+	declare -A generated_files=( ["${1}/.files_generated_oadm"]="${1}/.files_generated_oadm" \
+	                           ["${1}/.files_generated_oc"]="${1}/.files_generated_oc"     \
+	                           ["${1}/.files_generated_openshift"]="${1}/.files_generated_openshift" )
+
+	for generated_file in "${generated_files[@]}"; do
+	if [ -e ${generated_file} ]; then
```

Comment:
`-s` is a more appropriate check here

use `[[` instead of `[`

indent the block of the `for` loop

also -- if the file is empty, won't the `while read` just no-op? not sure why we need the `if` check here